tests/ui/needless_lifetimes.rs

```diff
@@ -259,4 +259,36 @@ mod issue4291 {
     }
 }
 
+mod nested_elision_sites {
```

Comment:
We could add the following tests:
* The snippet with the outlives bound that creates the error for the first case you identified, verbatim.
* Returning a function pointer that takes a function pointer as a parameter, with and without named lifetimes.
* Returning `impl Trait` (just one would suffice) that takes a function pointer as a parameter, with and without named lifetimes.
* Returning/Taking a boxed `dyn Trait`, with and without named lifetimes.
* A function that takes `impl Trait` in parameter position, with and without named lifetimes.

(NOTE: For the cases without named lifetimes, we should make it so it triggers the lint).

Reply:
> * The snippet with the outlives bound that creates the error for the first case you identified, verbatim.
The example in the issue
```
fn get_x<'a>(i32_pointer: &'a i32) -> impl Fn() -> &'a i32 {
    move || i32_pointer
}
```
is an equivalent of 
```rust
fn nested_fn_trait_bound<'a>(i: &'a i32) -> impl Fn() -> &'a i32 {
    move || i
}
```
Doesn't need to be verbatim imo.

> *  Returning a function pointer that takes a function pointer as a parameter, with and without named lifetimes. 
> * Returning impl Trait (just one would suffice) that takes a function pointer as a parameter, with and without named lifetimes.

I'll add cases so that fn pointers and trait bound have both with and without named lifetimes cases covered

> * Returning/Taking a boxed dyn Trait, with and without named lifetimes.

I haven't checked the behavior of dyn trait bounds (but the elision rules get more complex as I understood from the reference). I'll add a case and see how that goes.

Reply:
>Doesn't need to be verbatim imo.

Sorry, I was not very precise, I was refering to [the comment that opened the issue](https://github.com/rust-lang/rust-clippy/issues/2944#issue-343146515), specifically this snippet:
```rust
use std::collections::{HashMap, HashSet};

struct MyMap {
    inner: HashMap<u32, HashSet<String>>,
}

impl MyMap {
    fn iter<'a>(&'a self, key: u32) -> Option<impl Iterator<Item = impl AsRef<str> + 'a>> {
        self.inner.get(&key).map(|set| set.iter())
    }
}

fn main() {}
```

>I'll add cases so that fn pointers and trait bound have both with and without named lifetimes cases covered

I was refering to cases like these:
```rust
    fn fun1<'a>(_: &'a i32) -> fn(fn(&'a i32) -> &'a i32) -> i32 {
        |f| 42
    }
    fn fun2<'a>(_: &'a i32) -> fn(fn(&i32) -> &i32) -> i32 {
        |f| 42
    }

    fn fun3<'a>(_: &'a i32) -> impl Fn(fn(&'a i32)) {
        |f| ()
    }
    fn fun4<'a>(_: &'a i32) -> impl Fn(fn(&i32)) {
        |f| ()
    }
```

>I haven't checked the behavior of dyn trait bounds

I thought about these just for completeness. If it turns out the lint does not work as expected, we can open another issue to deal with that.

Reply:
I've added tests cases similar to those you suggest. 

Concerning the `Iterator<Item = impl AsRef<str> + 'a>>` mentioned in the issue, there is a similar one in `needless_lifetimes_impl_trait.rs`: 
```rust 
impl Bar {
    fn baz<'a>(&'a self) -> impl Foo + 'a {
        Baz { bar: self }
    }
}
```
I'll copy it in the main test file as well.

As for `dyn Fn`, I'm struggling to find a realistic example that doesn't break other elision rules such as multiple identical lifetime references in the fn arguments.

Reply:
>Concerning the Iterator<Item = impl AsRef<str> + 'a>> mentioned in the issue, there is a similar one in needless_lifetimes_impl_trait.rs

Yeah that one can be included too :+1: 
But the one I was refering in my previous comment should be included too, because the type  `Option<impl Iterator<Item = impl AsRef<str> + 'a>>` includes the outlives bound for the nested `impl Trait`, which should not be linted now.

>As for dyn Fn, I'm struggling to find a realistic example that doesn't break other elision rules such as multiple identical lifetime references in the fn arguments.

Let's leave that one out then, to be fair it's not really related to the issue.

Reply:
I'll try to dig into your rework ASAP, but if I can't make it today it may take a couple days as I will be out of town.

Reply:
> But the one I was refering in my previous comment should be included too, because the type Option<impl Iterator<Item = impl AsRef<str> + 'a>> includes the outlives bound for the nested impl Trait, which should not be linted now.

@ebroto As I understand it, nested `impl Trait`s with lifetimes are only problematic if it is itself a candidate for elision (Fn, FnOnce, FnMut). The example you mention works fine with elided lifetimes (`Iterator<Item = impl AsRef<str> + '_>>`). In the issue analysis, I propose to continue linting this case.

Reply:
>In the issue analysis, I propose to continue linting this case.

I would prefer not linting here or giving the final suggestion, as I think we should never suggest code that does not compile, but this is indeed debatable. I would like to have more opinions on this, so I've [asked in zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Suggest.20erroneous.20code.20if.20fix.20is.20simple.3F/near/210751560).

Reply:
After hearing others' thoughts on this, I think we can go with your proposal :+1: 

Following @flip1995 suggestion there, we could explain this special case in the documentation / known problems.